pkg/storage/engine/pebble_merge.go

```diff
@@ -0,0 +1,260 @@
+}
+
+// mergeTimeSeries combines two `InternalTimeSeriesData`s and returns the result as an
+// `InternalTimeSeriesData`.  The inputs cannot be merged if they have different start
+// timestamps or sample durations.
+func mergeTimeSeries(oldTs, newTs roachpb.InternalTimeSeriesData) (roachpb.InternalTimeSeriesData, error) {
+	if oldTs.StartTimestampNanos != newTs.StartTimestampNanos {
+		return roachpb.InternalTimeSeriesData{}, errors.Errorf("start timestamp mismatch")
+	}
+	if oldTs.SampleDurationNanos != newTs.SampleDurationNanos {
+		return roachpb.InternalTimeSeriesData{}, errors.Errorf("sample duration mismatch")
+	}
+
+	// TODO(ajkr): confirm it is the case that (1) today's CRDB always merges timeseries
+	// values in columnar format, and (2) today's CRDB does not need to be downgrade-
+	// compatible with any version that supports row format only. Then we can drop support
+	// for row format entirely. It requires significant cleanup effort as many tests target
+	// the row format.
+	if len(oldTs.Offset) > 0 || len(newTs.Offset) > 0 {
+		ensureColumnar(&oldTs)
+		ensureColumnar(&newTs)
+		proto.Merge(&oldTs, &newTs)
+		sortAndDeduplicateColumns(&oldTs)
+	} else {
+		proto.Merge(&oldTs, &newTs)
+		sortAndDeduplicateRows(&oldTs)
+	}
+	return oldTs, nil
+}
+
+// mergeTimeSeriesValues attempts to merge two values which contain
+// InternalTimeSeriesData messages.
+func mergeTimeSeriesValues(oldTsBytes, newTsBytes []byte) ([]byte, error) {
+	var oldTs, newTs, mergedTs roachpb.InternalTimeSeriesData
+	if err := protoutil.Unmarshal(oldTsBytes, &oldTs); err != nil {
+		return nil, errors.Errorf("corrupted old timeseries: %v", err)
+	}
+	if err := protoutil.Unmarshal(newTsBytes, &newTs); err != nil {
+		return nil, errors.Errorf("corrupted new timeseries: %v", err)
+	}
+
+	var err error
+	if mergedTs, err = mergeTimeSeries(oldTs, newTs); err != nil {
+		return nil, errors.Errorf("mergeTimeSeries: %v", err)
+	}
+
+	if res, err := protoutil.Marshal(&mergedTs); err != nil {
+		return nil, errors.Errorf("corrupted merged timeseries: %v", err)
+	} else {
+		return res, nil
+	}
+}
+
+// merge combines two serialized `MVCCMetadata`s and returns the result as a serialized
+// `MVCCMetadata`.
+//
+// Replay Advisory: Because merge commands pass through raft, it is possible
+// for merging values to be "replayed". Currently, the only actual use of
+// the merge system is for time series data, which is safe against replay;
+// however, this property is not general for all potential mergeable types.
+// If a future need arises to merge another type of data, replay protection
+// will likely need to be a consideration.
+func merge(key, oldValue, newValue, buf []byte, log pebble.Logger) []byte {
+	const (
+		checksumSize = 4
+		tagPos       = checksumSize
+		headerSize   = checksumSize + 1
+	)
+
+	var oldMeta, newMeta, mergedMeta enginepb.MVCCMetadata
+	if err := protoutil.Unmarshal(oldValue, &oldMeta); err != nil {
+		log.Fatalf("corrupted old mvcc: %v", err)
```

Comment:
The fatals are due to no current mechanism for returning error from `merge`. It may be worth considering adding a way, though I also think getting rid of fatals will be a much larger project.